pkg/results/error.go

```diff
@@ -0,0 +1,83 @@
+package results
+
+import (
+	"fmt"
+)
+
+// Error holds a message and a child, allowing for an error
+// The common use-case here will be to wrap errors from callsites:
+//
+// if err := doSomething(data); err != nil {
+//     return results.ForReason(results.ReasonFoo).WithError(err).Errorf("could not do something for data: %v", data)
+// }
+type Error struct {
+	reason  Reason
+	message string
+	wrapped error
+}
+
+// Error makes an Error an error
+func (e Error) Error() string {
+	return e.message
+}
+
+// FullReason provides the chain of error reasons, divided by colons
+func (e Error) FullReason() string {
+	reasonedError, ok := e.wrapped.(*Error)
```

Comment:
lets use `errors.As` here  (requires to implement `Unwrap` on the type). Also, some funcs have value and some have pointer receivers, please stick to one of the two